src/config.cpp

```diff
@@ -717,102 +840,41 @@ class SimulationConfig::Parser
-                                        debugStr));
-                    }
-                    parseOptional(valueIt, "variance", input.variance);
-                    break;
-                }
-                case Input::Module::shot_noise:
-                    parseMandatory(valueIt, "rise_time", debugStr, input.riseTime);
-                    parseMandatory(valueIt, "decay_time", debugStr, input.decayTime);
-                    parseOptional(valueIt,
-                                  "random_seed",
-                                  input.randomSeed,
-                                  {parseRun().randomSeed});
-                    parseOptional(valueIt, "dt", input.dt, {0.25});
-                    parseMandatory(valueIt, "rate", debugStr, input.rate);
-                    parseMandatory(valueIt, "amp_mean", debugStr, input.ampMean);
-                    parseMandatory(valueIt, "amp_var", debugStr, input.ampVar);
-                    break;
-                case Input::Module::relative_shot_noise:
-                    parseMandatory(valueIt, "rise_time", debugStr, input.riseTime);
-                    parseMandatory(valueIt, "decay_time", debugStr, input.decayTime);
-                    parseOptional(valueIt,
-                                  "random_seed",
-                                  input.randomSeed,
-                                  {parseRun().randomSeed});
-                    parseOptional(valueIt, "dt", input.dt, {0.25});
-                    parseMandatory(valueIt, "amp_cv", debugStr, input.ampCv);
-                    parseMandatory(valueIt, "mean_percent", debugStr, input.meanPercent);
-                    parseMandatory(valueIt, "sd_percent", debugStr, input.sdPercent);
-                    break;
-                case Input::Module::hyperpolarizing:
-                    break;
-                default:
-                    throw SonataError(moduledebugStr);
-                }
-                break;
-            case Input::InputType::spikes:
-                if (input.module == SimulationConfig::Input::Module::synapse_replay) {
-                    parseMandatory(valueIt, "spike_file", debugStr, input.spikeFile);
-                    input.spikeFile = toAbsolute(_basePath, input.spikeFile);
-                    parseOptional(valueIt, "source", input.source);
-                } else {
-                    throw SonataError(moduledebugStr);
-                }
+
+            InputBase::Module module;
+            parseMandatory(valueIt, "module", debugStr, module);
+
+            const auto input =
+                parseInputModule(valueIt, module, _basePath, parseRun().randomSeed, debugStr);
+            result[it.key()] = input;
+
+            auto inputType = nonstd::visit([](const auto& v) { return v.inputType; }, input);
+
+            switch (inputType) {
+            case InputBase::InputType::current_clamp: {
+                assert(nonstd::holds_alternative<SimulationConfig::InputLinear>(input) ||
```

Comment:
As I tested, `assert` here doesn't report error or abort the process. I don't know why. Perhaps better to throw with a clear message?

Reply:
> I don't know why.

Was it a debug build? the `assert()` macro is a no-op in release.

This is more for debug builds, and only to make sure that when someone adds a new variant, it gets added in the right place.

Reply:
ah, I thought it was for detecting the wrong combination of a (valid) module type and a (valid) input type from users. Do we have a check for that ?

Reply:
I don't think so, I've repurposed this test for that, and added a test.